### DISCUSSION TITLE
Undici.eventSource

### DISCUSSION THREAD

**Discussion by tushar32** [nodejs/undici#2976] [Q&A] [answered]
Is eventSource still experimental?  Can we use this feature?
I am using node 18 currently and it shows experimental warning.

**Comment by metcoder95** [comment]
You can use it, but be aware is still experimental and its subject to changes over the time, until it goes out of experimental

**Reply by tushar32** [reply]
Thanks for reply! How soon we can expect this feature to be production ready?

**Reply by metcoder95** [reply]
Depends on several factors; it is not possible to give timeline for it

**Reply by Uzlopak** [reply, marked as answer]
The expectation that it is "not" production ready is wrong. I am the main developer for this implementation. I can assure you that I was very thorough in the implementation. It is developed up to the spec.

But, there can be reasons, why you would e.g. use the "eventsource" npm package. E.g. the eventsource npm package allows you set headers while instantiating the eventsource connection. Or set a specific proxy - something which is also possible with undici, but not done on eventsource level but iirc via setting a globaldispatcher. 

Our EventSource implementation is now released for about 6-8 weeks. So we have not enough feedback to claim that it is battletested. But this will only be possible if we see some adaption by users like you and getting some feedback. 

Anyhow: the npm package "eventsource" is downloaded about 20 mio times per month and basically battletested. Our EventSource implementation is just an offer. Nobody forces you to use it ;). But it is implemented by the spec, same as the npm package "eventsource". So you should be able to replace the one with the other. So if you are worried about potential bugs or security issues, you still have the fallback option. 

My plan is to expose EventSource as a global in node 22.

Feedback is always nice. ;)

**Reply by tushar32** [reply]
Hey Aras, 

Thanks for you nice explaination and a suggestion. 

I will definately  try eventSource npm .

I used undici eventsource and I obeserve that whenever I use it I had to restart the server then it able to pick up the event data.

**Reply by Uzlopak** [reply]
@tushar32 that sounds like a bug? Can you pease provide a mvce?

**Reply by tushar32** [reply]
I will get back to you. let me do couple of more tests

**Reply by tushar32** [reply]
Hi @Uzlopak ,

I am using 
  Node version: v18.16.0
  "express": "^4.18.3"
  "nodemon": "^3.1.0"


<img width="472" alt="image" src="https://github.com/nodejs/undici/assets/30208305/0b98aff9-6164-4338-a65b-79616e6e13e3">
<img width="469" alt="image" src="https://github.com/nodejs/undici/assets/30208305/1977a95f-5689-47ac-aa7c-675ce51eaf78">

The issue is when I start the epxress js server it throws error initially but when I restart the nodemon (2nd sceenshot) it's starts working.

**Reply by Uzlopak** [reply]
@tushar32 
If there is a bug, then please create an issue and provide a mvce.

**Reply by tushar32** [reply]
Ok I will create one. 
Just a one question!

Does eventsource uses polling internally?

**Reply by Uzlopak** [reply]
No, it has a constant Connection
